source/jormungandr/jormungandr/scenarios/new_default.py

```diff
@@ -170,18 +175,220 @@ def tag_journeys(resp):
+
+    """
+    We want to find the solutions covers as many sections as possible which has less sections(less transfer to do)
+    """
+    the_best_idx = np.lexsort((nb_sections, integrity))[0]  # sort by integrity then by nb_sections
+
+    best_indexes = np.where(np.logical_and(nb_sections == nb_sections[the_best_idx],
+                                           integrity == integrity[the_best_idx]))[0]
+
+    logger.debug("Best Itegrity: {0}".format(integrity[the_best_idx]))
+    logger.debug("Best Nb sections: {0}".format(nb_sections[the_best_idx]))
+
+    return best_indexes, selection_matrix
+
+
 def culling_journeys(resp, request):
     """
-    remove some journeys if there are too many of them
-    since the journeys are sorted, we remove the last elt in the list
+    Remove some journeys if there are too many of them to have max_nb_journeys journeys.
+
+    The goal is to choose a bunch of journeys(max_nv_journeys) that covers as many as possible sections
+    but have as few as possible sum(sections)
+
+    Ex:
+
+    From:
+
+    Journey_1 : Line 1 -> Line 8 -> Bus 172
+    Journey_2 : Line 14 -> Line 6 -> Bus 165
+    Journey_3 : Line 14 -> Line 6 ->Line 8 -> Bus 165
 
-    no remove done in debug
+    W'd like to choose two journeys. The algo will return Journey_1 and Journey2.
+
+    Because
+    With Journey_1 and Journey_3, they cover all lines but have 5 transfers in all
+    With Journey_2 and Journey_3, they don't cover all lines(Line 1 is missing) and have 5 transfers in all
+
+    With Journey_1 and Journey_2, they cover all lines and have only 4 transfers in all -> OK
+
+    No removing done in debug
     """
+    logger = logging.getLogger(__name__)
+
     if request['debug']:
         return
 
-    if request["max_nb_journeys"] and len(resp.journeys) > request["max_nb_journeys"]:
-        del resp.journeys[request["max_nb_journeys"]:]
+    if not request["max_nb_journeys"] or request["max_nb_journeys"] >= len(resp.journeys):
+        logger.debug('No need to cull journeys')
+        return
+
+    logger.debug('Trying to culling the journeys')
+
+    """
+    To create a candidates pool, we choose only journeys that are NOT tagged as 'comfort' and 'best' and we create a
+    section set from that pool
+
+    Ex:
+    Journey_1 (Best): Line 14 -> Line 8 -> Bus 172
+    Journey_2 : Line 14 -> Line 6 -> Bus 165
+    Journey_3 : Line 14 -> Line 8 -> Bus 165
+
+    The candidate pool will be like [Journey_2, Journey_3]
+    The sections set will be like set([Line 14, Line 6, Line 8, Bus 165])
+    """
+    candidates_pool, sections_set = _build_candidate_pool_and_sections_set(resp)
+
+    nb_journeys_must_have = len(resp.journeys) - len(candidates_pool)
+    logger.debug("There are {0} journeys we must keep".format(nb_journeys_must_have))
+    nb_journeys_to_find = request["max_nb_journeys"] - nb_journeys_must_have
+    if nb_journeys_to_find == 0:
```

Comment:
What if nb_journeys_to_find is negative? For example, if max_nb_journeys == 1, we only want best